### ISSUE TITLE
renderModuleFactory does not reject when error is thrown in application

### ISSUE THREAD

**Issue by FrozenPandaz** [angular/angular#33642]
# 🐞 bug report

### Affected Package
`@angular/platform-server`

### Is this a regression?

No, I tested a v4 and v6 and the behavior is consistent

### Description

`renderModuleFactory` does not reject for errors thrown during rendering.

Examples:

* `@Component` `ngOnInit`
* Router `resolve`
* Probably more

## 🔬 Minimal Reproduction

https://github.com/FrozenPandaz/angular-bugs/tree/render-module-factory

```bash
npm i && npm run prerender
```

## 🔥 Exception or Error
<pre><code> jason@pop-os  ~/projects/temp/temp/angular-bugs   render-module-factory  npm run prerender

> ng8-app@0.0.0 prerender /home/jason/projects/temp/temp/angular-bugs
> ng build && ng run ng8-app:server && node ./main.js

Generating ES5 bundles for differential loading...
ES5 bundle generation complete.

chunk {runtime} runtime-es2015.js, runtime-es2015.js.map (runtime) 6.16 kB [entry] [rendered]
chunk {runtime} runtime-es5.js, runtime-es5.js.map (runtime) 6.16 kB [entry] [rendered]
chunk {styles} styles-es2015.js, styles-es2015.js.map (styles) 10.1 kB [initial] [rendered]
chunk {styles} styles-es5.js, styles-es5.js.map (styles) 14.2 kB [initial] [rendered]
chunk {main} main-es2015.js, main-es2015.js.map (main) 48.2 kB [initial] [rendered]
chunk {main} main-es5.js, main-es5.js.map (main) 55.3 kB [initial] [rendered]
chunk {polyfills-es5} polyfills-es5.js, polyfills-es5.js.map (polyfills-es5) 789 kB [initial] [rendered]
chunk {vendor} vendor-es2015.js, vendor-es2015.js.map (vendor) 3.75 MB [initial] [rendered]
chunk {vendor} vendor-es5.js, vendor-es5.js.map (vendor) 5.21 MB [initial] [rendered]
chunk {polyfills} polyfills-es2015.js, polyfills-es2015.js.map (polyfills) 264 kB [initial] [rendered]
Date: 2019-11-07T04:43:04.373Z - Hash: c69b66b49129c074781b - Time: 13236ms
Hash: 2a979d82dfc31ad8eb0a
Time: 2896ms
Built at: 11/06/2019 11:43:08 PM
      Asset      Size  Chunks             Chunk Names
    main.js    65 KiB    main  [emitted]  main
main.js.map  36.2 KiB    main  [emitted]  main
Entrypoint main = main.js main.js.map
chunk {main} main.js, main.js.map (main) 54.8 KiB [entry] [rendered]
ERROR Error: hi
    at AppComponent.ngOnInit (/home/jason/projects/temp/temp/angular-bugs/dist/ng8-app-server/main.js:233:15)
    at checkAndUpdateDirectiveInline (/home/jason/projects/temp/temp/angular-bugs/node_modules/@angular/core/bundles/core.umd.js:21240:23)
    at checkAndUpdateNodeInline (/home/jason/projects/temp/temp/angular-bugs/node_modules/@angular/core/bundles/core.umd.js:29610:24)
    at checkAndUpdateNode (/home/jason/projects/temp/temp/angular-bugs/node_modules/@angular/core/bundles/core.umd.js:29572:20)
    at prodCheckAndUpdateNode (/home/jason/projects/temp/temp/angular-bugs/node_modules/@angular/core/bundles/core.umd.js:30113:9)
    at Object.updateDirectives (/home/jason/projects/temp/temp/angular-bugs/dist/ng8-app-server/main.js:187:264)
    at Object.updateDirectives (/home/jason/projects/temp/temp/angular-bugs/node_modules/@angular/core/bundles/core.umd.js:29901:76)
    at Object.checkAndUpdateView (/home/jason/projects/temp/temp/angular-bugs/node_modules/@angular/core/bundles/core.umd.js:29554:18)
    at ViewRef_.detectChanges (/home/jason/projects/temp/temp/angular-bugs/node_modules/@angular/core/bundles/core.umd.js:20829:26)
    at ApplicationRef.tick (/home/jason/projects/temp/temp/angular-bugs/node_modules/@angular/core/bundles/core.umd.js:27224:30)
promise resolved
 jason@pop-os  ~/projects/temp/temp/angular-bugs   render-module-factory  

</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
 ✘ jason@pop-os  ~/projects/temp/fs-datatable   master  ng v

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 7.3.1
Node: 10.16.3
OS: linux x64
Angular: 7.2.15
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.13.9
@angular-devkit/build-angular     0.13.9
@angular-devkit/build-optimizer   0.13.9
@angular-devkit/build-webpack     0.13.9
@angular-devkit/core              7.3.9
@angular-devkit/schematics        7.3.1
@angular/cdk                      7.3.7
@angular/cli                      7.3.1
@angular/material                 7.3.7
@ngtools/webpack                  7.3.9
@schematics/angular               7.3.1
@schematics/update                0.13.1
rxjs                              6.3.3
typescript                        3.2.2
webpack                           4.29.0
</code></pre>

**Comment by FrozenPandaz** [comment]
cc @vikerman

**Comment by arturovt** [comment]
@FrozenPandaz 

I have looked into your reproduction. This exception is caught during the very first change detection. `tick` is called inside the `ApplicationRef.prototype._loadComponent`. `tick` wraps `detectChanges` in the `try-catch` statement and uses the default `ErrorHandler`, that just logs this exception to console.

The very first thought that came to my mind is just to override the default `ErrorHandler` with your custom one on the server:
```ts
@Injectable()
export class ServerErrorHandler implements ErrorHandler {
  handleError(error: Error): void {
    AppServerModule.error$.next(error);
  }
}

@NgModule({
  imports: [AppModule, ServerModule],
  bootstrap: [AppComponent],
  providers: [
    {
      provide: ErrorHandler,
      useClass: ServerErrorHandler
    }
  ]
})
export class AppServerModule {
  static error$ = new Subject<Error>();
}
```
And do something like:
```js
AppServerModuleNgFactory.moduleType.error$.subscribe(error => {
  console.log('error = ', error);
});

renderModuleFactory(AppServerModuleNgFactory, {
  document: index
}).then(() => {
  console.log('promise resolved');
});
```

---

P.S. I really do not like this code and I'm sure there could have been some better solution. This is a quick fix...

**Comment by FrozenPandaz** [comment]
Thank you so much! Will give that a try. We probably don't want the default to swallow the error though... Do we?

**Comment by Platonn** [comment]
IMHO those are criteria of an ideal solution: (What do you think @alan-agius4?)
- in custom Angular app: developers should be able to conveniently decide which types of errors they want to treat as "critical"
- in Angular/Universal framework: the framework should propagate those critical errors (with via any medium they like) from the Angular app to the Express app
- in custom ExpressJS app: information about those critical errors should be available, returned from the ngExpressEngine in a standard way, familiar to ExpressJS developers, e.g. as a callback `(err,html)`.

**Comment by Platonn** [comment]
I believe there is an architectural/business decision to be made: should errors be propagated **immediately** when they occur VS. **lately** only after the app is stable and the rendering finished?

### Errors propagated immediately:
#### Pros:
- the client can get an instant response as soon as an error happens. no unnecessary waiting
#### Cons:
- after the error is emitted to Express (and possibly to a client), the rendering is not finished yet(!), meaning that there are still some async tasks. Those tasks can cause side-effects, e.g. make new outgoing http calls, print to logs, write to files(?). It might be confusing that `ngExpressEngine` emitted an error. When observing the application activity in the monitoring tools, tt might be confusing that an "Errored render" is a "zombie" that can still perform side-effects. 

### Errors propagated lately (only after app is stable):
#### Pros:
- intuitive. "error emitted" means "render is finished"
#### Cons:
- client unnecessarily waits for the render to be completed, even though we already know a critical error occured and we could potentially respond to client's request immediately

I have hard time to choose. My heart says the "late" errors propagation is better, because it's 100% safe and predictable. But my brain tells me it's not as performant as could be.

Just curious, which option you're more fond of, @alan-agius4 ?

**Comment by Platonn** [comment]
By the way, when we agree on the final architectural design, I'm happy to contribute with a PR, if it works for you. The topic of Angular SSR [really interests me](https://dev.to/krisplatis/how-angular-ssr-works-under-the-hood-source-code-analysis-amj).

**Comment by Platonn** [comment]
Hi @alan-agius4, sorry for bugging you. I'm aware you have lot on your plate.
Is it on your roadmap to consider fixing this issue or would you mind if I create a PR?

PS. the duplicate of this issue has 23 likes 👍

**Comment by Platonn** [comment]
Hi @alan-agius4 , thanks for your reply. 

Could you help me understand please, which aspects do you perceive as not trivial?
- not changing the shape of existing Public API (i.e. avoid breaking changes)?
- architecturally deciding on _when_ to propagate errors (immediately VS. only after app `isStable`)?
- ... or maybe you meant something else?

If there is anything more you think I should give a special attention, I'm happy to adapt.

Thanks in advance!

**Comment by webberig** [comment]
I ended up here while investigating an issue on our production website. A page was returning a 200 OK response in a corrupted state. I noticed that the `ng-state` was missing in the response.

The server logs mentioned following error:
```
   Ignoring BEFORE_APP_SERIALIZED Exception:  TypeError: Converting circular structure to JSON
       --> starting at object with constructor 'r'
       |     property 'upload' -> object with constructor 'b'
       --- property '_request' closes the circle
       at JSON.stringify (<anonymous>)
       at e.toJson (file:///app/dist/apps/web/server/chunk-7FR65NHR.mjs:5:31328)
```

Due to an application error, there was an unserializable object being put in the transferstate object. This caused the error above. Unfortunately, the error is being swallowed so the page returns a 200 OK without the transferstate object, which in turn caused some issues for our website client-side.

The code containing the error handling is here:
https://github.com/angular/angular/blob/f35b2ef47cef778e35c3f62ba51212b3585a33f2/packages/platform-server/src/utils.ts#L207

Ideally we have a way of handling this kind of error and produce an error page instead.

**Comment by alan-agius4** [comment]
@webberig, your issue seems something slightly different from the original issue. Do you mind opening an new issue? Thanks.

**Comment by zola-dev** [comment]
Our report ([angular/angular-cli#33522](https://github.com/angular/angular-cli/issues/33522)) was closed as a duplicate of this issue. It is on the current pipeline: @angular/build prerendering with @angular/ssr on v22, where renderModuleFactory (this issue's title) no longer exists.

The outcome is the same as described here: 
a component render error is routed to the default ErrorHandler during ApplicationRef.tick and only logged, whenStable() resolves after it, and the build exits 0 shipping the degraded page.

We opened it as a separate issue following the suggestion in this thread (Dec 2025) to open a new one for a related report:
https://github.com/angular/angular/issues/33642#issuecomment-3602029155
It has a fresh v22 reproduction and a proposed mechanism (exposing rethrowApplicationErrors, which TestBed already has, for server/prerender rendering).

Could the new report be reopened as its own tracker for the current pipeline? If this issue is meant to cover it as well, glad to keep it here.
